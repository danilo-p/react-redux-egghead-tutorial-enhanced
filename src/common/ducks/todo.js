// Initial State
export const INITIAL_STATE = {};

// Actions
export const ADD = 'ADD_TODO';
export const TOGGLE = 'TOGGLE_TODO';

// Reducer
export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
  case ADD:
    return {
      id: action.id,
      text: action.text,
      completed: false,
    };
  case TOGGLE:
    if (state.id !== action.id) {
      return state;
    }
    return {
      ...state,
      completed: !state.completed,
    };
  default:
    return state;
  }
};

export const createTodoId = text => Date.now() + text;

// Action Creators
export function addTodo(text, id = createTodoId(text)) {
  return {
    type: ADD,
    id,
    text,
  };
}

export const toggleTodo = id => ({
  type: TOGGLE,
  id,
});
