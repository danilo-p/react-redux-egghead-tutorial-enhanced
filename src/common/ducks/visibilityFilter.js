/**
 * The "visibilityFilter" duck.
 *
 * @module common/ducks/visibilityFilter
 */

/* ======================================
=            Initial State              =
====================================== */

/** "visibilityFilter" duck initial state */
export const INITIAL_STATE = 'SHOW_ALL';

/* ======  End of Initial State  ====== */


/* =====================================
=            Action Types              =
===================================== */

/** SET_VISIBILITY_FILTER action type */
export const SET = 'SET_VISIBILITY_FILTER';

/* ======  End of Action Types  ====== */


/* ================================
=            Reducer              =
================================ */

/**
 * The "visibilityFilter" duck reducer
 *
 * @memberof module:common/ducks/visibilityFilter
 * @param {Object} [state=INITIAL_STATE] The current state
 * @param {Object} action The action to be handled
 * @returns {Object} The new state
 */
export default function reducer(state = INITIAL_STATE, action) {
  switch (action.type) {
  case SET:
    return action.filter;
  default:
    return state;
  }
}

/* ======  End of Reducer  ====== */


/* ========================================
=            Action Creators              =
======================================== */

/**
 * Creates a new SET_VISIBILITY_FILTER action.
 *
 * @memberof module:common/ducks/visibilityFilter
 * @param {String} filter The visibility filter
 * @returns {Object} The new SET_VISIBILITY_FILTER action
 */
export const setVisibilityFilter = filter => ({
  type: 'SET_VISIBILITY_FILTER',
  filter,
});

/* ======  End of Action Creators  ====== */
