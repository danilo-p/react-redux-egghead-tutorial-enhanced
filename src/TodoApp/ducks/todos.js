import todo, {
  ADD,
  TOGGLE,
} from './todo';

// Actions
export const REMOVE = 'REMOVE_TODO';
export { ADD, TOGGLE };

// Reducer
export default (
  state = [
    {
      id: 0,
      text: 'test0',
      completed: false,
    },
    {
      id: 1,
      text: 'test1',
      completed: false,
    },
    {
      id: 2,
      text: 'test2',
      completed: false,
    },
  ],
  action,
) => {
  switch (action.type) {
  case ADD:
    return [
      ...state,
      todo(undefined, action),
    ];
  case TOGGLE:
    return state.map(t => todo(t, action));
  case REMOVE:
    return []; // TODO: implement the remove todo
  default:
    return state;
  }
};
