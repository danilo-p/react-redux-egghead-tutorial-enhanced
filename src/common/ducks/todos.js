import todo, {
  ADD,
  TOGGLE,
} from './todo';

// Actions
export const REMOVE = 'REMOVE_TODO';
export { ADD, TOGGLE };

// Reducer
export default (
  state = [],
  action,
) => {
  switch (action.type) {
  case ADD:
    return [
      ...state,
      todo(undefined, action),
    ];
  case TOGGLE:
    return state.map(t => todo(t, action));
  case REMOVE:
    return []; // TODO: implement the remove todo
  default:
    return state;
  }
};
