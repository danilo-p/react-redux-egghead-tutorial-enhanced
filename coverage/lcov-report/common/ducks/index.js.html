<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for common/ducks/index.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> / <a href="index.html">common/ducks</a> index.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>0/0</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">/**
 * Module that bundles all the ducks in one reducer. It imports all the ducks
 * and combine their reducers with the combineReducers method of the redux
 * library.
 *
 * @module common/ducks
 * @author Danilo Pimentel
 */
&nbsp;
import { combineReducers } from 'redux';
&nbsp;
import todo from './todo';
import todos from './todos';
import visibilityFilter from './visibilityFilter';
&nbsp;
// Here all the reducers are combined into a single one reducer that is
// exported as default and used to create the application store.
export default combineReducers({
  todo,
  todos,
  visibilityFilter,
});
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Fri Dec 29 2017 17:51:47 GMT-0200 (-02)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
