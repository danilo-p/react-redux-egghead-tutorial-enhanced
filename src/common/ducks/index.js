/**
 * Module that bundles all the ducks in one reducer. It imports all the ducks
 * and combine their reducers with the combineReducers method of the redux
 * library.
 *
 * @module common/ducks
 */

import { combineReducers } from 'redux';

import todo from './todo';
import todos from './todos';
import visibilityFilter from './visibilityFilter';

// Here all the reducers are combined into a single one reducer that is
// exported as default and used to create the application store.
export default combineReducers({
  todo,
  todos,
  visibilityFilter,
});
