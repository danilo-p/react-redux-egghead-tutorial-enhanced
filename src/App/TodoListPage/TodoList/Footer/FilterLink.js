import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Button } from 'semantic-ui-react';
import { setVisibilityFilter } from 'common/ducks/visibilityFilter';

/**
 * Describes a link that is used to filter the todos on the app.
 *
 * @param {Object} props The component properties
 * @param {Boolean} props.active Is the filter that the link describes active?
 * @param {Array} props.children The link children components
 * @param {Function} props.onClick Button click handler
 * @returns {jsx} The link template
 */
const FilterLink = ({
  active,
  children,
  onClick,
}) => (
  <Button
    secondary
    disabled={active}
    size="mini"
    onClick={onClick}
  >
    {children}
  </Button>
);

FilterLink.propTypes = {
  active: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func.isRequired,
};

const mapStateToProps = (
  state,
  containerProps,
) => ({
  active: containerProps.filter === state.visibilityFilter,
});

const mapDispatchToProps = (
  dispatch,
  containerProps,
) => ({
  onClick: () => {
    dispatch(setVisibilityFilter(containerProps.filter));
  },
});

/**
 * Exports the container component for the FilterLink
 */
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(FilterLink);
