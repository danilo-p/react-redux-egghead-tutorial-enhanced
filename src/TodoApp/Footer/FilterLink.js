import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Button } from 'semantic-ui-react';
import { setVisibilityFilter } from '../ducks/visibilityFilter';

const Link = ({
  active,
  children,
  onClick,
}) => (
  <Button
    secondary
    disabled={active}
    size="mini"
    onClick={onClick}
  >
    {children}
  </Button>
);

Link.propTypes = {
  active: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func.isRequired,
};

const mapStateToProps = (
  state,
  ownProps,
) => ({
  active: ownProps.filter === state.visibilityFilter,
});

const mapDispatchToProps = (
  dispatch,
  ownProps,
) => ({
  onClick: () => {
    dispatch(setVisibilityFilter(ownProps.filter));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Link);
