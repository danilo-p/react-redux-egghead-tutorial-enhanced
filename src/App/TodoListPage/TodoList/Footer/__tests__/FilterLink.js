import React from 'react';
import { INITIAL_STATE } from 'common/ducks';
import { setVisibilityFilter } from 'common/ducks/visibilityFilter';
import { Button } from 'semantic-ui-react';
import configureMockStore from 'redux-mock-store';
import FilterLink from '../FilterLink';

const mockStore = configureMockStore();

describe('FilterLink', () => {
  let filterLink = null;
  let store = null;

  const filter = 'TEST_FILTER';
  const children = ['Test filter'];

  beforeEach(() => {
    store = mockStore(INITIAL_STATE);

    filterLink = (
      <FilterLink
        filter={filter}
        store={store}
      >
        {children}
      </FilterLink>
    );

    filterLink = mount(filterLink);
  });

  it('should render the component', () => {
    expect(filterLink).toMatchSnapshot();
  });

  it('should always render one Button', () => {
    const buttons = filterLink.find(Button);
    expect(buttons.length).toBe(1);
  });

  describe('the rendered Button', () => {
    let button = null;

    beforeEach(() => {
      button = filterLink.find(Button).first();
    });

    it('should render the passed children', () => {
      expect(button.children().length).toBe(children.length);
    });

    it('should dispatch a SET_VISIBILITY_FILTER action on click', () => {
      button.simulate('click');

      const actions = store.getActions();
      expect(actions.length).toBe(1);

      const setVisibilityFilterAction = actions[0];
      expect(setVisibilityFilterAction).toEqual(setVisibilityFilter(filter));
    });
  });
});
