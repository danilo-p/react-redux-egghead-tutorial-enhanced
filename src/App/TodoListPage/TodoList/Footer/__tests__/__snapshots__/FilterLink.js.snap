// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterLink should render the component 1`] = `
<Connect(FilterLink)
  filter="TEST_FILTER"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <FilterLink
    active={false}
    filter="TEST_FILTER"
    onClick={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  >
    <Button
      as="button"
      disabled={false}
      onClick={[Function]}
      secondary={true}
      size="mini"
    >
      <button
        className="ui mini secondary button"
        disabled={undefined}
        onClick={[Function]}
        role="button"
        tabIndex={undefined}
      >
        Test filter
      </button>
    </Button>
  </FilterLink>
</Connect(FilterLink)>
`;
